### ISSUE TITLE
[BUG] Identical blocks might break dependecy_graph 

### ISSUE THREAD

**Issue by gwwatkin** [latticesurgery-com/lattice-surgery-compiler#176]
Issue Description
-----------------

This line suggests it:

https://github.com/latticesurgery-com/lattice-surgery-compiler/blob/faad0347ff9f1d8a2b8d6fe2f6d9ca3055e82866/src/lsqecc/logical_lattice_ops/dependency_graph.py#L134

If two blocks are identical they will be interpreted as the same key in this map which doesn't sounds like the desired behavior.

Proposed Solution
-----------------

Need to investigate if this is the case and find a solution. Also the code in `dependency_graph.py` is very hard to read because it's missing type annotations. Especially because it's generic/templated code it should be annotated.

Additional References
---------------------
The equivalence of rotations
https://github.com/latticesurgery-com/lattice-surgery-compiler/blob/faad0347ff9f1d8a2b8d6fe2f6d9ca3055e82866/src/lsqecc/pauli_rotations/rotation.py#L172-L177

and measurements
https://github.com/latticesurgery-com/lattice-surgery-compiler/blob/faad0347ff9f1d8a2b8d6fe2f6d9ca3055e82866/src/lsqecc/pauli_rotations/rotation.py#L211-L216

**Comment by alexnguyenn** [comment]
A quick solution I can think of is to remove that method since `DependencyGraph.Node` has parents and children list, which is enough for us to figure out the adjacency list anyway.

**Comment by gwwatkin** [comment]
It should, as we are only ever using it in a debug file to view the graph. I think working on this class is lower priority than anything in the [UF epic](https://github.com/latticesurgery-com/general-issue-tracker/issues/3), so I think it's fine to leave this for now, but good to have the issue logged if we do start to use it.

**Comment by gwwatkin** [comment]
If we go back to the plan of parallelizing operations

**Comment by alexnguyenn** [comment]
I think it will also break this: 
https://github.com/latticesurgery-com/lattice-surgery-compiler/blob/faad0347ff9f1d8a2b8d6fe2f6d9ca3055e82866/src/lsqecc/logical_lattice_ops/dependency_graph.py#L113-L116

Which is part of the DAG generating method, with `visited` being a set. Looking through the [codecov report](https://app.codecov.io/gh/latticesurgery-com/lattice-surgery-compiler/blob/dev/src/lsqecc/logical_lattice_ops/dependency_graph.py) seems like it is not covered by our test case which is why I missed it. I might need to look more into it but I think we can just use the node there instead of op (by `visisted.add(parent)`)?

**Comment by alexnguyenn** [comment]
Yeah I will fix the type annotations in there as well. They are a mess 😓

**Comment by alexnguyenn** [comment]
https://github.com/latticesurgery-com/lattice-surgery-compiler/blob/faad0347ff9f1d8a2b8d6fe2f6d9ca3055e82866/src/lsqecc/logical_lattice_ops/dependency_graph.py#L41-L45 

lol I override `__eq__` for Node as well. Might need to get rid of it.

**Comment by gwwatkin** [comment]
It's good we have all these notes. Don't think we need to act on it yet.
